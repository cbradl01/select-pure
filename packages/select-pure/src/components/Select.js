import { css, LitElement, html } from "lit-element";

import { KEYS } from "./constants";

export class Select extends LitElement {
  static get styles() {
    return css`
      .select-wrapper {
        position: relative;
      }
      .select-wrapper:hover .select {
        z-index: 2;
      }
      .select {
        bottom: 0;
        display: flex;
        flex-wrap: wrap;
        left: 0;
        position: absolute;
        right: 0;
        top: 0;
        width: var(--select-width, 100%);
      }
      .label:focus {
        outline: var(--select-outline, 2px solid #e3e3e3);
      }
      .label:after {
        border-bottom: 1px solid var(--color, #000);
        border-right: 1px solid var(--color, #000);
        box-sizing: border-box;
        content: "";
        display: block;
        height: 10px;
        margin-top: -2px;
        transform: rotate(45deg);
        transition: 0.2s ease-in-out;
        width: 10px;
      }
      .label.visible:after {
        margin-bottom: -4px;
        margin-top: 0;
        transform: rotate(225deg);
      }
      select {
        -webkit-appearance: none;
        -moz-appearance: none;
        appearance: none;
        position: relative;
        opacity: 0;
        z-index: 1;
      }
      select,
      .label {
        align-items: center;
        background-color: var(--background-color, #fff);
        border-radius: var(--border-radius, 4px);
        border: var(--border-width, 1px) solid var(--border-color, #000);
        box-sizing: border-box;
        color: var(--color, #000);
        cursor: pointer;
        display: flex;
        font-family: var(--font-family, inherit);
        font-size: var(--font-size, 14px);
        font-weight: var(--font-weight, 400);
        height: var(--select-height, 44px);
        justify-content: space-between;
        padding: var(--padding, 0 10px);
        width: 100%;
      }
      .dropdown {
        background-color: var(--border-color, #000);
        border-radius: var(--border-radius, 4px);
        border: var(--border-width, 1px) solid var(--border-color, #000);
        display: none;
        flex-direction: column;
        gap: var(--border-width, 1px);
        justify-content: space-between;
        max-height: calc(var(--select-height, 44px) * 4 + var(--border-width, 1px) * 3);
        overflow-y: scroll;
        position: absolute;
        top: calc(var(--select-height, 44px) + var(--dropdown-gap, 0px));
        width: calc(100% - var(--border-width, 1px) * 2);
        z-index: var(--dropdown-z-index, 2);
      }
      .dropdown.visible {
        display: flex;
      }
      .disabled {
        background-color: var(--disabled-background-color, #bdc3c7);
        color: var(--disabled-color, #ecf0f1);
        cursor: default;
      }
    `;
  }

  static get properties() {
    return {
      options: {
        type: Array,
      },
      visible: {
        type: Boolean,
      },
      selectedOption: {
        type: Object,
      },
      disabled: {
        type: Boolean,
      },
      value: {
        type: String,
      },
      name: {
        type: String,
      },
      formName: {
        type: String,
      },
    };
  }

  constructor() {
    super();

    // bindings
    this.close = this.close.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.processOptions = this.processOptions.bind(this);
    this.watchNativeSelect = this.watchNativeSelect.bind(this);
    this.processForm = this.processForm.bind(this);

    // properties
    this.options = [];
    this.visible = false;
    this.selectedOption = {};
    this.disabled = this.getAttribute("disabled") !== null;
    this.name = this.getAttribute("name");
    this.id = this.getAttribute("id");
    this.formName = this.name || this.id;
    this.value = null;
  }

  firstUpdated() {
    this.processOptions();
    this.watchNativeSelect();
    // :( Fails with React
    // window.addEventListener("DOMContentLoaded", this.processForm);
  }

  // public

  get selectedIndex() {
    return this.nativeSelect.selectedIndex;
  }

  set selectedIndex(index) {
    this.onSelect(this.options[index].value);
  }

  open() {
    if (this.disabled) {
      return;
    }
    document.body.removeEventListener("click", this.close);
    this.visible = true;
    setTimeout(() => {
      // :( don't close the select right away
      document.body.addEventListener("click", this.close);
    });
  }

  close() {
    this.visible = false;
    document.body.removeEventListener("click", this.close);
  }

  enable() {
    this.disabled = false;
  }

  disable() {
    this.disabled = true;
  }

  // private methods

  processForm() {
    this.form = this.closest("form");
    if (!this.form) {
      return;
    }
    this.hiddenInput = document.createElement("input");
    this.hiddenInput.setAttribute("type", "hidden");
    this.hiddenInput.setAttribute("name", this.formName);
    this.form.appendChild(this.hiddenInput);
  }

  watchNativeSelect() {
    this.nativeSelect.addEventListener("change", () => {
      this.selectedIndex = this.nativeSelect.selectedIndex;
    });
  }

  processOptions() {
    this.nativeSelect = this.shadowRoot.querySelector("select");
    const options = this.querySelectorAll("option-pure");
    for (let i = 0; i < options.length; i++) {
      const { value, label, select, unselect, selected, hidden, disabled } = options[i].getOption();
      this.options.push({
        label,
        value,
        select,
        unselect,
        hidden,
        disabled,
      });
      if (selected) {
        this.selectedOption = options[i];
        this.nativeSelect.selectedIndex = i;
      }
      options[i].onSelect = this.onSelect;

      if (i === options.length - 1 && !this.selectedOption.value) {
        this.selectedOption = options[0];
        options[0].select();
        this.nativeSelect.selectedIndex = i;
      }
    }
  }

  onSelect(optionValue) {
    for (let i = 0; i < this.options.length; i++) {
      const option = this.options[i];
      if (option.value === optionValue) {
        this.selectOption(option, i);
        continue;
      }
      option.unselect();
    }
    if (this.form) {
      this.hiddenInput.value = this.value;
      const event = new Event("change", { bubbles: true });
      this.hiddenInput.dispatchEvent(event);
    }
    this.visible = false;
  }

  selectOption(option, index) {
    this.selectedOption = option;
    this.value = option.value;
    option.select();
    this.nativeSelect.selectedIndex = index;
    this.afterSelect();
  }

  afterSelect() {
    this.dispatchEvent(new Event("change"));
  }

  renderOptions() {
    return this.options.map(({ value, label, hidden, disabled }) => {
      const isSelected = this.selectedOption.value === value;
      return html`
        <option
          value=${value}
          ?selected=${isSelected}
          ?hidden=${hidden}
          ?disabled=${disabled}
        >
          ${label}
        </option>
      `;
    });
  }

  handleKeyPress(event) {
    if (event.which === KEYS.ENTER || event.which === KEYS.TAB) {
      this.open();
    }
  }

  render() {
    const labelClassNames = ["label"];
    if (this.disabled) {
      labelClassNames.push("disabled");
    }
    if (this.visible) {
      labelClassNames.push("visible");
    }
    return html`
      <div class="select-wrapper">
        <select ?disabled=${this.disabled} name="${this.name}" id=${this.id}>
          ${this.renderOptions()}
        </select>

        <div class="select">
          <div
            class="${labelClassNames.join(" ")}"
            @click="${this.visible ? this.close : this.open}"
            @keydown="${this.handleKeyPress}"
            tabindex="0"
          >
            ${this.selectedOption.label}
          </div>
          <div class="dropdown${this.visible ? " visible" : ""}">
            <slot></slot>
          </div>
        </div>
      </div>
    `;
  }
}
