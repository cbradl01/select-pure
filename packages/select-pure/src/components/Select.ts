import { boundMethod } from "autobind-decorator";
import { LitElement, html } from "lit";
import { ifDefined } from "lit-html/directives/if-defined.js";
import { customElement } from "lit/decorators/custom-element.js";
import { property } from "lit/decorators/property.js";
import { query } from "lit/decorators/query.js";

import { KEYS, defaultOption } from "./../constants";
import { Option, OptionPureElement } from "./../models";
import { selectStyles } from "./../styles";

@customElement("select-pure")
export class SelectPure extends LitElement {
  static get styles() {
    return selectStyles;
  }

  @property() options: Option[] = [];
  @property() visible: boolean = false;
  @property() selectedOption: Option = defaultOption;
  @property() _selectedOptions: Option[] = [];
  @property() disabled: boolean = false;
  @property() isMultipleSelect: boolean = false;
  @property() name: string = "";
  @property() _id: string = "";
  @property() formName: string = "";
  @property() value: string = "";
  @property() values: string[] = [];
  @property() defaultLabel: string = "";
  @property() totalRenderedChildOptions: number = -1;
  @query("select") nativeSelect!: HTMLSelectElement;

  private form: HTMLFormElement | null = null;
  private hiddenInput: HTMLInputElement | null = null;


  connectedCallback() {
    super.connectedCallback();

    this.disabled = this.getAttribute("disabled") !== null;
    this.isMultipleSelect = this.getAttribute("multiple") !== null;
    this.name = this.getAttribute("name") || "";
    this._id = this.getAttribute("id") || "";
    this.formName = this.name || this.id;
    this.defaultLabel = this.getAttribute("default-label") || "";
  }

  public open() {
    if (this.disabled) {
      return;
    }
    this.visible = true;
    this.removeEventListeners();
    document.body.addEventListener("click", this.close, true);
  }

  @boundMethod
  public close(event?: Event) {
    // @ts-ignore
    if (event && this.contains(event.target)) {
      return;
    }
    this.visible = false;
    this.removeEventListeners();
  }

  public enable() {
    this.disabled = false;
  }

  public disable() {
    this.disabled = true;
  }

  get selectedIndex(): number | undefined {
    return this.nativeSelect?.selectedIndex;
  }

  set selectedIndex(newSelectedIndex: number | undefined) {
    if (!newSelectedIndex && newSelectedIndex !== 0) {
      return;
    }
    this.selectOptionByValue(this.options[newSelectedIndex].value);
  }

  get selectedOptions() {
    return this.nativeSelect?.selectedOptions;
  }

  render() {
    const labelClassNames = ["label"];
    if (this.disabled) {
      labelClassNames.push("disabled");
    }
    if (this.visible) {
      labelClassNames.push("visible");
    }

    return html`
      <div class="select-wrapper">
        <select
          @change=${this.handleNativeSelectChange}
          ?disabled=${this.disabled}
          ?multiple=${this.isMultipleSelect}
          name="${ifDefined(this.name || undefined)}"
          id=${ifDefined(this.id || undefined)}
          size="1"
        >
          ${this.getNativeOptionsHtml()}
        </select>
        <div class="select">
          <div
            class="${labelClassNames.join(" ")}"
            @click="${this.visible ? this.close : this.open}"
            @keydown="${this.openDropdownIfProperKeyIsPressed}"
            tabindex="0"
          >
            ${this.getDisplayedLabel()}
          </div>
          <div class="dropdown${this.visible ? " visible" : ""}">
            <slot @slotchange=${this.initializeSelect}></slot>
          </div>
        </div>
      </div>
    `;
  }

  private handleNativeSelectChange() {
    this.selectedIndex = this.nativeSelect?.selectedIndex;
  }

  private getNativeOptionsHtml() {
    return this.options.map(this.getSingleNativeOptionHtml);
  }

  @boundMethod
  private getSingleNativeOptionHtml({ value, label, hidden, disabled }: Option) {
    return html`
      <option
        value=${value}
        ?selected=${this.isOptionSelected(value)}
        ?hidden=${hidden}
        ?disabled=${disabled}
      >
        ${label}
      </option>
    `;
  }

  private isOptionSelected(value: String): boolean {
    let isOptionSelected = this.selectedOption.value === value;
    if (this.isMultipleSelect) {
      isOptionSelected = Boolean(this._selectedOptions.find(option => option.value === value));
    }
    return isOptionSelected;
  }

  private openDropdownIfProperKeyIsPressed(event: KeyboardEvent) {
    if (event.key === KEYS.ENTER || event.key === KEYS.TAB) {
      this.open();
    }
  }

  private getDisplayedLabel() {
    if (this.isMultipleSelect && this._selectedOptions.length) {
      return this.getMultiSelectLabelHtml();
    }

    return this.selectedOption.label || this.defaultLabel;
  }

  @boundMethod
  private getMultiSelectLabelHtml() {
    if (this._selectedOptions.length>1){
      return this._selectedOptions.length + " choices selected"
    } else {
      return html`
        <div class="multi-selected-wrapper">
          ${this._selectedOptions.map(this.getMultiSelectSelectedOptionHtml)}
        </div>
      `;
    }
  }

  @boundMethod
  private getMultiSelectSelectedOptionHtml({ label, value }: Option) {
    return html`
//       <span class="multi-selected">
      <span>
        ${label}
//         <span
//           class="cross"
//           @click=${(event: Event) => this.fireOnSelectCallback(event, value)}
//         >
//         </span>
      </span>
    `;
  }

  private fireOnSelectCallback(event: Event, value: string) {
    event.stopPropagation();
    this.selectOptionByValue(value);
  }

  @boundMethod
  private initializeSelect() {
    this.processChildOptions();
    this.selectDefaultOptionIfNoneSelected();
    this.appendHiddenInputToClosestForm();
  }

  private processChildOptions() {
    const options = this.querySelectorAll("option-pure");
    this.totalRenderedChildOptions = options.length;
    for (let i = 0; i < options.length; i++) {
      this.initializeSingleOption(options[i] as OptionPureElement, i);
    }
  }

  private selectDefaultOptionIfNoneSelected() {
    const shouldSelectDefaultOption = !this.selectedOption.value && !this.isMultipleSelect && this.options.length;
    if (shouldSelectDefaultOption) {
      this.selectOptionByValue(this.options[0].value);
    }
  }

  @boundMethod
  private initializeSingleOption(optionElement: OptionPureElement, optionIndex: number) {
    optionElement.setOnSelectCallback(this.selectOptionByValue);
    this.options[optionIndex] = optionElement.getOption();
    if (this.options[optionIndex].selected) {
      this.selectOptionByValue(this.options[optionIndex].value);
    }
  }

  @boundMethod
  private removeEventListeners() {
    document.body.removeEventListener("click", this.close);
  }

  @boundMethod
  private appendHiddenInputToClosestForm() {
    this.form = this.closest("form");
    if (!this.form || this.hiddenInput) {
      return;
    }
    this.hiddenInput = document.createElement("input");
    this.hiddenInput.setAttribute("type", "hidden");
    this.hiddenInput.setAttribute("name", this.formName);
    this.form.appendChild(this.hiddenInput);
  }

  private unselectAllOptions() {
    for (let i = 0; i < this.options.length; i++) {
      this.options[i].unselect();
    }
  }

  @boundMethod
  private selectOptionByValue(newOptionValue: string) {
    const option = this.options.find(({ value }) => value === newOptionValue);
    if (!option) {
      return;
    }
    this.setSelectValue(option);
  }

  private setSelectValue(optionToBeSelected: Option) {
    if (this.isMultipleSelect) {
      this.setMultiSelectValue(optionToBeSelected);
    } else {
      this.setSingleSelectValue(optionToBeSelected);
    }
    this.updateHiddenInputInForm();
    this.dispatchChangeEvent();
  }

  private dispatchChangeEvent() {
    this.dispatchEvent(new Event("change"));
  }

  private setMultiSelectValue(optionToBeSelected: Option) {
    const indexInSelectedOptions = this._selectedOptions.indexOf(optionToBeSelected);
    const isAlreadySelected = indexInSelectedOptions !== -1;
    if (isAlreadySelected) {
      this.values.splice(indexInSelectedOptions, 1);
      this._selectedOptions.splice(indexInSelectedOptions, 1);
      optionToBeSelected.unselect();
    } else {
      this.values.push(optionToBeSelected.value);
      this._selectedOptions.push(optionToBeSelected);
      optionToBeSelected.select();
    }
    this.requestUpdate();
  }

  private setSingleSelectValue(optionToBeSelected: Option) {
    this.unselectAllOptions();
    this.close();
    this.selectedOption = optionToBeSelected;
    this.value = optionToBeSelected.value;
    optionToBeSelected.select();
  }

  private updateHiddenInputInForm() {
    if (!this.form || !this.hiddenInput) {
      return;
    }
    this.hiddenInput.value = this.isMultipleSelect ? this.values.join(",") : this.value;
    const event = new Event("change", { bubbles: true });
    this.hiddenInput.dispatchEvent(event);
  }
}
