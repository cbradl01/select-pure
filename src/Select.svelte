<svelte:options tag="select-pure" />

<p>test</p>
