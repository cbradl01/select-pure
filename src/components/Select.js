import { css, LitElement, html } from "lit-element";

export class Select extends LitElement {
  static get styles() {
    return css`
      select {
        display: none;
      }
      .select {
        display: flex;
        flex-wrap: wrap;
        height: 44px;
        position: relative;
        width: 400px;
      }
      .selected {
        align-items: center;
        border-radius: 4px;
        border: 1px solid black;
        cursor: pointer;
        display: flex;
        height: 100%;
        justify-content: space-between;
        padding: 0 10px;
        width: 100%;
      }
      .dropdown {
        background-color: black;
        border-radius: 4px;
        border: 1px solid black;
        display: none;
        flex-direction: column;
        gap: 0.5px;
        justify-content: space-between;
        overflow: hidden;
        position: absolute;
        top: 44px;
        width: calc(100% - 2px);
        z-index: 2;
      }
      .dropdown.visible {
        display: flex;
      }
    `;
  }

  static get properties() {
    return {
      options: {
        type: Array,
      },
      visible: {
        type: Boolean,
      },
      selectedOption: {
        type: Object,
      },
    };
  }

  constructor() {
    super();

    // bindings
    this.addOption = this.addOption.bind(this);
    this.close = this.close.bind(this);
    this.onSelect = this.onSelect.bind(this);

    // properties
    this.options = [];
    this.visible = false;
    this.selectedOption = {};


    const options = this.querySelectorAll("option-pure");
    for (let i = 0; i < options.length; i++) {
      options[i].onInit = this.addOption;
      options[i].onSelect = this.onSelect;
    }
  }

  addOption(option) {
    const { value, label, select, unselect, selected } = option;
    this.options.push({
      label,
      value,
      select,
      unselect,
    });
    if (selected) {
      this.selectedOption = option;
    }
  }

  onSelect(optionValue) {
    for (let i = 0; i < this.options.length; i++) {
      const option = this.options[i];
      if (option.value === optionValue) {
        this.selectedOption = option;
        option.select();
        continue;
      }
      option.unselect();
    }
  }

  get selectedIndex() {
    return 0;
  }

  renderOptions() {
    return this.options.map(({ value, label }) => html`<option value=${value}>${label}</option>`);
  }

  open(event) {
    event.stopPropagation(); // causes problem when one select is open and other is clicked
    this.visible = true;
    document.body.addEventListener("click", this.close);
  }

  close() {
    this.visible = false;
    document.body.removeEventListener("click", this.close);
  }

  render() {
    return html`
      <div class="select-wrapper">
        <select>
          ${this.renderOptions()}
        </select>

        <div class="select" aria-hidden="true">
          <div class="selected" @click="${this.visible ? this.close : this.open}">
            ${this.selectedOption.label || "Default label"}
          </div>
          <div class="dropdown${this.visible ? " visible" : ""}">
            <slot></slot>
          </div>
        </div>
      </div>
    `;
  }
}
