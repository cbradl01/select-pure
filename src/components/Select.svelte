<svelte:options tag="select-pure" />

<script>
  import { selected } from "./../store.js";
  let label;
  let dropdownVisible = false;
  let select;

  const unsubscribeValue = selected.subscribe(value => {
    // callback
	});

  function toggleDropdown() {
    dropdownVisible = !dropdownVisible;
    const options = select.parentNode.querySelectorAll("option-pure");
    console.log(select);
    console.log(options);
    console.log($$slots);
  }

  function onChange() {
    console.log("ON CHANGE!");
  }
</script>

<div class="select" on:click={toggleDropdown} bind:this={select}>
  {#if $selected.value}
    <span>{$selected.label}</span>
  {/if}
  <select
    class:visible="{dropdownVisible}"
    class="dropdown"
  >
    <slot></slot>
  </select>
</div>

<style>
  .select {
    align-items: center;
    background: #f9f9f8;
    border-radius: 4px;
    border: 1px solid rgba(0, 0, 0, 0.15);
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.04);
    box-sizing: border-box;
    color: #363b3e;
    cursor: pointer;
    display: flex;
    font-size: 16px;
    font-weight: 500;
    justify-content: left;
    min-height: 44px;
    padding: 5px 10px;
    position: relative;
    transition: 0.2s;
    width: 100%;
  }
  .dropdown {
    border-radius: 4px;
    border: 1px solid rgba(0, 0, 0, 0.15);
    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.04);
    box-sizing: border-box;
    color: #363b3e;
    display: none;
    left: 0;
    max-height: 221px;
    overflow-y: scroll;
    position: absolute;
    top: 50px;
    width: 100%;
    z-index: 5;
  }
  .visible {
    display: block;
  }
</style>
