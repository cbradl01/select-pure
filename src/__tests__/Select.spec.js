import { render, fireEvent } from "@testing-library/svelte";

import Select from "./../Select.svelte";

describe("Select component", () => {
  it("renders empty select", () => {
    const { container } = render(Select, {
      name: "select",
    });
    expect(container).toMatchSnapshot();
  });

  it("opens dropdown", async() => {
    const { container } = render(Select, {
      name: "select",
    });

    const selectWrapper = document.querySelector(".select");
    await fireEvent.click(selectWrapper);

    expect(container).toMatchSnapshot();
  });

  it("closes dropdown", async() => {
    const { container } = render(Select, {
      name: "select",
    });

    const selectWrapper = document.querySelector(".select");

    await fireEvent.click(selectWrapper);
    await fireEvent.click(selectWrapper);

    expect(container).toMatchSnapshot();
  });

  it("closes dropdown on click outside", async() => {
    const { container } = render(Select, {
      name: "select",
    });

    const selectWrapper = document.querySelector(".select");

    await fireEvent.click(selectWrapper);
    await fireEvent.click(document.body);

    expect(container).toMatchSnapshot();
  });
});

