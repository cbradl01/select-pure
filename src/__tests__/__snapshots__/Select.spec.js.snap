// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component closes dropdown 1`] = `
<body>
  <div>
    <div
      class="select"
    >
       
      <div
        class="dropdown"
      />
    </div>
  </div>
</body>
`;

exports[`Select component closes dropdown on click outside 1`] = `
<body>
  <div>
    <div
      class="select"
    >
       
      <div
        class="dropdown visible"
      />
    </div>
  </div>
</body>
`;

exports[`Select component opens dropdown 1`] = `
<body>
  <div>
    <div
      class="select"
    >
       
      <div
        class="dropdown visible"
      />
    </div>
  </div>
</body>
`;

exports[`Select component renders empty select 1`] = `
<body>
  <div>
    <div
      class="select"
    >
       
      <div
        class="dropdown"
      />
    </div>
  </div>
</body>
`;
